src/thor/triplegbuilder.cc

```diff
@@ -91,93 +91,190 @@ void AssignAdmins(const AttributesController& controller,
+    found->set_end_shape_index(index);
+  } // We are starting a new incident
+  else {
+    auto* new_incident = leg.mutable_incidents()->Add();
+    new_incident->set_id(incident->id);
+    new_incident->set_begin_shape_index(index);
+    new_incident->set_end_shape_index(index);
+    // TODO: copy the rest of the metadata
+  }
+}
+
+/**
+ * Chops up the shape for an edge so that we have shape points where speeds change along the edge
+ * and where incidents occur along the edge. Also sets the various per shape point attributes
+ * such as time, distance, speed. Also updates the incidents list on the edge with their shape indices
+ * @param controller
+ * @param tile
+ * @param edge
+ * @param shape
+ * @param shape_begin
+ * @param leg
+ * @param src_pct
+ * @param tgt_pct
+ * @param edge_seconds
+ * @param cut_for_traffic
+ * @param incidents
+ */
 void SetShapeAttributes(const AttributesController& controller,
                         const GraphTile* tile,
                         const DirectedEdge* edge,
                         std::vector<PointLL>& shape,
                         size_t shape_begin,
-                        TripLeg& trip_path,
+                        TripLeg& leg,
                         double src_pct,
                         double tgt_pct,
                         double edge_seconds,
-                        bool cut_for_traffic) {
+                        bool cut_for_traffic,
+                        const std::vector<baldr::GraphReader::Incident>& incidents) {
   // TODO: if this is a transit edge then the costing will throw
 
-  if (trip_path.has_shape_attributes()) {
-    // A list of percent along the edge and corresponding speed (meters per second)
-    std::vector<std::tuple<double, double>> speeds;
-    double speed = (edge->length() * (tgt_pct - src_pct)) / edge_seconds;
-    if (cut_for_traffic) {
-      // TODO: we'd like to use the speed from traffic here but because there are synchronization
-      // problems with those records changing between when we used them to make the path and when we
-      // try to grab them again here, we instead rely on the total time from PathInfo and just do the
-      // cutting for now
-      const auto& traffic_speed = tile->trafficspeed(edge);
-      if (traffic_speed.breakpoint1 > 0) {
-        speeds.emplace_back(traffic_speed.breakpoint1 / 255.0, speed);
-        if (traffic_speed.breakpoint2 > 0) {
-          speeds.emplace_back(traffic_speed.breakpoint2 / 255.0, speed);
-          if (traffic_speed.speed3 != UNKNOWN_TRAFFIC_SPEED_RAW) {
-            speeds.emplace_back(1, speed);
-          }
+  // bail if nothing to do
+  if (!cut_for_traffic && incidents.empty() &&
+      !controller.category_attribute_enabled(kShapeAttributesCategory))
+    return;
+
+  // initialize shape_attributes once
+  if (!leg.has_shape_attributes() &&
+      controller.category_attribute_enabled(kShapeAttributesCategory)) {
+    leg.mutable_shape_attributes();
+  }
```

Comment:
this got refactored a bit. previously inside of triplegbuilder::build it was initializing shape attributes if that was enabled on the controller, now we do it once here. actually, i guess we dont even need to check if it has it anymore first since protobuf does that check internally.. derp